### PR TITLE
Bugz-536 Bot player

### PR DESCRIPTION
**Pull request by MiladNazeri**
https://highfidelity.atlassian.net/browse/BUGZ-536

### CODE REVIEW

File: usefulUtilities/botPlayer/assignmentClientPlayer.js [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+        }
+
+        if (message.uuid !== scriptUUID) {
+            return;
+        }
+
+        switch (message.action){
+            case "PLAY":
+                if (!player.isPlaying()) {
+                    var randoInterval = Math.max(Math.floor(Math.random() * INTERVAL_TIME), 7500);
```
Comment (MiladNazeri) [inline usefulUtilities/botPlayer/assignmentClientPlayer.js]: This section is what we were playing with.  Probably could be finessed a bit better.

File: usefulUtilities/botPlayer/README.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
+# Bot Player
+Bot Player helps with domain testing bots that need to be placed in a certain location, are a certain unique set, and that need to last through domain resets.
+
+## Setup
+1. use the botRecorder.js to record the bots that you need.  When you click on the tablet button, it will start recording and then stop after you click it again. 
+2. Keep the script update up to check on the status and to see where the file was saved.
```
Comment (zfox23) [inline usefulUtilities/botPlayer/README.md]: what is the "script update"?

File: usefulUtilities/botPlayer/README.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+3. Upload your bot files to a s3 directory(this is currently hardcoded in the manager file)
+4. Add the following two scripts to the Domain scripts [Place here when ready].
```
Comment (zfox23) [inline usefulUtilities/botPlayer/README.md]: what's "place here when ready" ?

File: usefulUtilities/botPlayer/README.md (hunk 2 above)
Comment (MiladNazeri) [reply usefulUtilities/botPlayer/README.md]: ahh, I most have zoned out while writing that.  I meant the script HMD log lol.

File: usefulUtilities/botPlayer/README.md (hunk 3 above)
Comment (MiladNazeri) [reply usefulUtilities/botPlayer/README.md]: that's for me to update when I have actual release files.  I can make them now though.

File: usefulUtilities/botPlayer/README.md (hunk 2 above)
Comment (MiladNazeri) [reply usefulUtilities/botPlayer/README.md]: done

File: usefulUtilities/botPlayer/README.md (hunk 3 above)
Comment (MiladNazeri) [reply usefulUtilities/botPlayer/README.md]: done

File: usefulUtilities/botPlayer/assignmentClientManager.js [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+            var player = availableAssignmentClientPlayers[botCount];
+            player.play();
+            botCount++;
+
+            if (botCount >= botsFound + 1) {
+                return;
+            }
+        }
+
+        Script.setTimeout(function() {
```
Comment (zfox23) [inline usefulUtilities/botPlayer/assignmentClientManager.js]: why is this necessary?

File: usefulUtilities/botPlayer/assignmentClientManager.js [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+            var currentBotUrl = BASE_PATH + "AVATAR_TEST" + i + ".hfr";
+            request(currentBotUrl, function(error){
+                if (error) {
+                    botRecordingFound = false;
+                } else {
+                    botsFound++;
+                    botList.push(currentBotUrl);
+                }
+            });
+            if (!botRecordingFound) {
```
Comment (zfox23) [inline usefulUtilities/botPlayer/assignmentClientManager.js]: doesn't this have to be inside the `request()` ?
EDIT: nevermind, request is synchronous. could you please change the name of the function to something like `requestSync` so that it's more clear?

Comment (zfox23) [reply usefulUtilities/botPlayer/assignmentClientManager.js]: also, what's supposed to happen in the `break` case?

File: usefulUtilities/botPlayer/assignmentClientManager.js [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+    var ASSIGNMENT_CLIENT_MESSANGER_CHANNEL = "ASSIGNMENT_CLIENT_MESSANGER_CHANNEL";
+
+    // Array of the assignment client players and their assignment client player object
+    var availableAssignmentClientPlayers = [];
+
+    // Current playing bot count we are at
+    var botCount = 0;
+
+    // Current registered bount count
+    var botRegisterdCount = 0;
```
Comment (zfox23) [inline usefulUtilities/botPlayer/assignmentClientManager.js]:
```suggestion
    var botRegisteredCount = 0;
```

File: usefulUtilities/botPlayer/assignmentClientManager.js [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+                availableAssignmentClientPlayers.push( 
+                    new AssignmentClientPlayerObject(message.uuid, fileName));
+                botRegisterdCount++;
+                var messageToSend = JSON.stringify({
+                    action: "AC_AVAILABLE_UPDATE",
+                    newAvailableACs: availableAssignmentClientPlayers.length
+                });
+                Messages.sendMessage(ASSIGNMENT_CLIENT_MESSANGER_CHANNEL, messageToSend);
+                break;
+            case "ARE_YOU_THERE_MANAGER_ITS_ME_BOT":
```
Comment (zfox23) [inline usefulUtilities/botPlayer/assignmentClientManager.js]: 😂 amazing

File: usefulUtilities/botPlayer/assignmentClientManager.js [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+                Messages.sendMessage(ASSIGNMENT_MANAGER_CHANNEL, JSON.stringify({
+                    action: "REGISTER_MANAGER",
+                    uuid: sender
+                }));
+                break;
+            default:
+                console.log("unrecongized action in assignmentClientManger.js");
```
Comment (zfox23) [inline usefulUtilities/botPlayer/assignmentClientManager.js]:
```suggestion
                console.log("unrecognized action in assignmentClientManger.js");
```

File: usefulUtilities/botPlayer/assignmentClientPlayer.js [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+    // *************************************
+    // #region MESSAGES
+    
+
+    // Handle messages fromt he manager
+    var IGNORE_PLAYER_MESSAGES = ["REGISTER_ME", "ARE_YOU_THERE_MANAGER"];
+    var playInterval = null;
+    var INTERVAL_TIME = 20000;
+    var MINIMUM_PLAY_TIME = 7500;
+    var randomScaler = 1.5
```
Comment (zfox23) [inline usefulUtilities/botPlayer/assignmentClientPlayer.js]:
```suggestion
    var randomScaler = 1.5;
```

Comment (zfox23) [reply usefulUtilities/botPlayer/assignmentClientPlayer.js]: also, what is this?

File: usefulUtilities/botPlayer/assignmentClientPlayer.js [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+
+        if (message.uuid !== scriptUUID) {
+            return;
+        }
+
+        switch (message.action){
+            case "PLAY":
+                if (!player.isPlaying()) {
+                    var randomPlayInterval = Math.max(Math.floor(Math.random() * INTERVAL_TIME), MINIMUM_PLAY_TIME);
+                    playInterval = Script.setInterval(function(){
```
Comment (zfox23) [inline usefulUtilities/botPlayer/assignmentClientPlayer.js]: you should make sure that you clear the old `playInterval` if one exists! Also, why does this interval exist in the first place? :D

File: usefulUtilities/botPlayer/assignmentClientManager.js (hunk 4 above)
Comment (MiladNazeri) [reply usefulUtilities/botPlayer/assignmentClientManager.js]: This is checking to see if all there are enough assignment clients for how many bots are needed.  Since there is no way to find out how many of this script are available, we need to keep polling to see if the number of bots needed === the number of available assignment clients

File: usefulUtilities/botPlayer/assignmentClientManager.js (hunk 5 above)
Comment (MiladNazeri) [reply usefulUtilities/botPlayer/assignmentClientManager.js]: First point - good idea, will change

second one - since this is sync, it breaks out of searching for more avatars needed and then the next function called starts the sequence

File: usefulUtilities/botPlayer/assignmentClientManager.js (hunk 6 above)
Comment (MiladNazeri) [reply usefulUtilities/botPlayer/assignmentClientManager.js]: thank you!

File: usefulUtilities/botPlayer/assignmentClientManager.js (hunk 8 above)
Comment (MiladNazeri) [reply usefulUtilities/botPlayer/assignmentClientManager.js]: thank you again!

File: usefulUtilities/botPlayer/assignmentClientPlayer.js (hunk 9 above)
Comment (MiladNazeri) [reply usefulUtilities/botPlayer/assignmentClientPlayer.js]: removing all those for now, but it was just to make the interval last a little longer than the timeout before it played again.

File: usefulUtilities/botPlayer/assignmentClientPlayer.js (hunk 10 above)
Comment (MiladNazeri) [reply usefulUtilities/botPlayer/assignmentClientPlayer.js]: Son of a...

Comment (MiladNazeri) [reply usefulUtilities/botPlayer/assignmentClientPlayer.js]: oh, to your question it's related to the last comment
